File: packages/contract/src/browse.ts
/** / — library browsing and item detail contracts. */

import { z } from "zod";
import {
  AudioTrackInfo as MediaFileAudioTrackInfo,
  SubtitleTrackInfo as MediaFileSubtitleTrackInfo,
} from "./media-files.js";

export const MediaKind = z.enum(["MOVIE", "SERIES", "SEASON", "EPISODE"]);
export const ContentProfile = z.enum(["GENERAL", "ANIME"]);
export const RelationType = z.enum([
  "MAIN",
  "MOVIE",
  "OVA",
  "SPECIAL",
  "RECAP",
  "SIDE_STORY",
  "PREQUEL",
  "SEQUEL",
]);

export const LibrarySummary = z.object({
  id: z.string(),
  name: z.string(),
  contentProfile: ContentProfile,
  mediaKinds: z.array(MediaKind),
});
export type LibrarySummary = z.infer<typeof LibrarySummary>;

export const MediaCard = z.object({
  id: z.string(),
  kind: MediaKind,
  title: z.string(),
  sortTitle: z.string(),
  year: z.number().int().nullable(),
  posterUrl: z.string().nullable(),
  backdropUrl: z.string().nullable(),
  /** Primary playable file, if any (leaf MOVIE/EPISODE items only) — null for SERIES/SEASON. */
  mediaFileId: z.string().nullable(),
  /** False for a bare folder item (empty dir) with no episodes/files — "not downloaded" in the UI. */
  isDownloaded: z.boolean(),
  genres: z.array(z.string()),
  createdAt: z.coerce.date(),
});
export type MediaCard = z.infer<typeof MediaCard>;

export const LibraryItemsParams = z.object({ id: z.string() });

export const MediaItemDetailParams = z.object({ id: z.string() });

/** Optional profileId scopes the detail response's watch data (episode watched marks, resume positions, item summary). */
export const MediaItemDetailQuery = z.object({ profileId: z.string().optional() });

export const CollectionEntry = z.object({
  relationType: RelationType,
  anchor: z.string().nullable(),
  item: MediaCard,
});

export const EpisodeCard = MediaCard.extend({
  seasonNumber: z.number().int().nullable(),
  episodeNumber: z.number().int().nullable(),
  runtimeMs: z.number().int().nullable(),
  /** Whether the requesting profile has watched this episode to completion. */
  watched: z.boolean().default(false),
  /** Resume position for the requesting profile — 0 when never started. */
  positionMs: z.number().default(0),
});
export type EpisodeCard = z.infer<typeof EpisodeCard>;

/** Per-profile watch summary for a media item (from PlaybackState). */
export const MediaItemWatch = z.object({
  watched: z.boolean(),
  positionMs: z.number(),
  durationMs: z.number().nullable(),
  /** Rewatch count — times watched to completion. */
  playCount: z.number(),
  lastWatchedAt: z.coerce.date().nullable(),
});
export type MediaItemWatch = z.infer<typeof MediaItemWatch>;

export const AudioTrackInfo = z.object({ streamIndex: z.number().int(), lang: z.string().nullable() });
export type AudioTrackInfo = z.infer<typeof AudioTrackInfo>;

export const MediaItemDetail = MediaCard.extend({
  overview: z.string().nullable(),
  originalTitle: z.string().nullable(),
  /** 0–10 normalized aggregate from the matched provider (TVmaze average, AniList averageScore/10, Jikan score). */
  rating: z.number().nullable(),
  genres: z.array(z.string()),
  studio: z.string().nullable(),
  children: z.array(MediaCard),
  /** Flattened grandchildren (episodes across all seasons) — empty unless kind === SERIES. */
  episodes: z.array(EpisodeCard),
  /** Show-scoped movies parented to this series (direct MOVIE children + legacy season-grandchild movies) — empty unless kind === SERIES. */
  movies: z.array(EpisodeCard),
  /** Primary file's audio streams — empty for SERIES/SEASON (no file of their own). */
  audioTracks: z.array(AudioTrackInfo),
  /** Video bitrate in kbps — primary file for MOVIE/EPISODE, mean across episode files for SERIES. Null when unprobed. */
  bitrateKbps: z.number().int().nullable(),
  /** Watch summary for the requesting profile — null when no profileId was passed. */
  watch: MediaItemWatch.nullable().default(null),
  /** Identity rows — what this item is currently matched to, for the "fix match" UI. */
  externalIds: z.array(z.object({ provider: z.string(), providerId: z.string() })),
  collections: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      kind: z.enum(["FRANCHISE", "MOVIE_SET"]),
      posterUrl: z.string().nullable(),
      relationType: RelationType,
      entries: z.array(CollectionEntry),
    }),
  ),
});
export type MediaItemDetail = z.infer<typeof MediaItemDetail>;

export const NotFoundError = z.object({ error: z.string() });

/** One playable file of a media item — everything a download/version picker needs. */
export const MediaFileDescriptor = z.object({
  mediaFileId: z.string(),
  /** The file browse/detail's `mediaFileId` points at — the primary playable. */
  isPrimary: z.boolean(),
  container: z.string().nullable(),
  durationMs: z.number().int().nullable(),
  sizeBytes: z.number().nullable(),
  bitrate: z.number().int().nullable(),
  video: z
    .object({
      codec: z.string().nullable(),
      width: z.number().int().nullable(),
      height: z.number().int().nullable(),
      frameRate: z.number().nullable(),
      isHdr: z.boolean(),
    })
    .nullable(),
  audioTracks: z.array(MediaFileAudioTrackInfo),
  subtitleTracks: z.array(MediaFileSubtitleTrackInfo),
});
export type MediaFileDescriptor = z.infer<typeof MediaFileDescriptor>;

export const MediaItemFilesParams = z.object({ id: z.string() });

export const MediaItemFilesResponse = z.array(MediaFileDescriptor);

File: native/flutter/lib/core/api/models/browse.dart
/// Mirrors packages/contract/src/browse.ts. Hand-written (no codegen) against
/// the frozen contract — see packages/contract/generated/openapi.json for the
/// source of truth if these drift.
library;

class LibrarySummary {
  final String id;
  final String name;
  final String contentProfile;
  final List<String> mediaKinds;

  LibrarySummary({required this.id, required this.name, required this.contentProfile, required this.mediaKinds});

  factory LibrarySummary.fromJson(Map<String, dynamic> j) => LibrarySummary(
        id: j['id'] as String,
        name: j['name'] as String,
        contentProfile: j['contentProfile'] as String,
        mediaKinds: (j['mediaKinds'] as List).cast<String>(),
      );
}

class MediaCard {
  final String id;
  final String kind; // MOVIE | SERIES | SEASON | EPISODE
  final String title;
  final String sortTitle;
  final int? year;
  final String? posterUrl;
  final String? backdropUrl;
  final String? mediaFileId;
  final bool isDownloaded;
  final List<String> genres;
  final DateTime? createdAt;

  MediaCard({
    required this.id,
    required this.kind,
    required this.title,
    required this.sortTitle,
    required this.year,
    required this.posterUrl,
    required this.backdropUrl,
    required this.mediaFileId,
    required this.isDownloaded,
    required this.genres,
    required this.createdAt,
  });

  factory MediaCard.fromJson(Map<String, dynamic> j) => MediaCard(
        id: j['id'] as String,
        kind: j['kind'] as String,
        title: j['title'] as String,
        sortTitle: j['sortTitle'] as String,
        year: j['year'] as int?,
        posterUrl: j['posterUrl'] as String?,
        backdropUrl: j['backdropUrl'] as String?,
        mediaFileId: j['mediaFileId'] as String?,
        isDownloaded: j['isDownloaded'] as bool? ?? false,
        genres: (j['genres'] as List? ?? const []).cast<String>(),
        createdAt: j['createdAt'] != null ? DateTime.tryParse(j['createdAt'] as String) : null,
      );
}

class EpisodeCard extends MediaCard {
  final int? seasonNumber;
  final int? episodeNumber;
  final int? runtimeMs;
  final bool watched;
  final int positionMs;

  EpisodeCard({
    required super.id,
    required super.kind,
    required super.title,
    required super.sortTitle,
    required super.year,
    required super.posterUrl,
    required super.backdropUrl,
    required super.mediaFileId,
    required super.isDownloaded,
    required super.genres,
    required super.createdAt,
    required this.seasonNumber,
    required this.episodeNumber,
    required this.runtimeMs,
    required this.watched,
    required this.positionMs,
  });

  factory EpisodeCard.fromJson(Map<String, dynamic> j) {
    final card = MediaCard.fromJson(j);
    return EpisodeCard(
      id: card.id,
      kind: card.kind,
      title: card.title,
      sortTitle: card.sortTitle,
      year: card.year,
      posterUrl: card.posterUrl,
      backdropUrl: card.backdropUrl,
      mediaFileId: card.mediaFileId,
      isDownloaded: card.isDownloaded,
      genres: card.genres,
      createdAt: card.createdAt,
      seasonNumber: j['seasonNumber'] as int?,
      episodeNumber: j['episodeNumber'] as int?,
      runtimeMs: j['runtimeMs'] as int?,
      watched: j['watched'] as bool? ?? false,
      positionMs: j['positionMs'] as int? ?? 0,
    );
  }
}

class MediaItemWatch {
  final bool watched;
  final int positionMs;
  final int? durationMs;
  final int playCount;
  final DateTime? lastWatchedAt;

  MediaItemWatch({
    required this.watched,
    required this.positionMs,
    required this.durationMs,
    required this.playCount,
    required this.lastWatchedAt,
  });

  factory MediaItemWatch.fromJson(Map<String, dynamic> j) => MediaItemWatch(
        watched: j['watched'] as bool,
        positionMs: j['positionMs'] as int,
        durationMs: j['durationMs'] as int?,
        playCount: j['playCount'] as int,
        lastWatchedAt: j['lastWatchedAt'] != null ? DateTime.tryParse(j['lastWatchedAt'] as String) : null,
      );
}

class DetailAudioTrack {
  final int streamIndex;
  final String? lang;
  DetailAudioTrack({required this.streamIndex, required this.lang});
  factory DetailAudioTrack.fromJson(Map<String, dynamic> j) =>
      DetailAudioTrack(streamIndex: j['streamIndex'] as int, lang: j['lang'] as String?);
}

class CollectionEntry {
  final String relationType;
  final String? anchor;
  final MediaCard item;
  CollectionEntry({required this.relationType, required this.anchor, required this.item});
  factory CollectionEntry.fromJson(Map<String, dynamic> j) => CollectionEntry(
        relationType: j['relationType'] as String,
        anchor: j['anchor'] as String?,
        item: MediaCard.fromJson(j['item'] as Map<String, dynamic>),
      );
}

class MediaCollection {
  final String id;
  final String name;
  final String kind; // FRANCHISE | MOVIE_SET
  final String? posterUrl;
  final String relationType;
  final List<CollectionEntry> entries;
  MediaCollection({
    required this.id,
    required this.name,
    required this.kind,
    required this.posterUrl,
    required this.relationType,
    required this.entries,
  });
  factory MediaCollection.fromJson(Map<String, dynamic> j) => MediaCollection(
        id: j['id'] as String,
        name: j['name'] as String,
        kind: j['kind'] as String,
        posterUrl: j['posterUrl'] as String?,
        relationType: j['relationType'] as String,
        entries: (j['entries'] as List? ?? const [])
            .map((e) => CollectionEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class MediaItemDetail extends MediaCard {
  final String? overview;
  final String? originalTitle;
  final double? rating;
  final String? studio;
  final List<MediaCard> children;
  final List<EpisodeCard> episodes;
  final List<EpisodeCard> movies;
  final List<DetailAudioTrack> audioTracks;
  final MediaItemWatch? watch;
  final List<MediaCollection> collections;

  MediaItemDetail({
    required super.id,
    required super.kind,
    required super.title,
    required super.sortTitle,
    required super.year,
    required super.posterUrl,
    required super.backdropUrl,
    required super.mediaFileId,
    required super.isDownloaded,
    required super.genres,
    required super.createdAt,
    required this.overview,
    required this.originalTitle,
    required this.rating,
    required this.studio,
    required this.children,
    required this.episodes,
    required this.movies,
    required this.audioTracks,
    required this.watch,
    required this.collections,
  });

  factory MediaItemDetail.fromJson(Map<String, dynamic> j) {
    final card = MediaCard.fromJson(j);
    return MediaItemDetail(
      id: card.id,
      kind: card.kind,
      title: card.title,
      sortTitle: card.sortTitle,
      year: card.year,
      posterUrl: card.posterUrl,
      backdropUrl: card.backdropUrl,
      mediaFileId: card.mediaFileId,
      isDownloaded: card.isDownloaded,
      genres: card.genres,
      createdAt: card.createdAt,
      overview: j['overview'] as String?,
      originalTitle: j['originalTitle'] as String?,
      rating: (j['rating'] as num?)?.toDouble(),
      studio: j['studio'] as String?,
      children: (j['children'] as List? ?? const []).map((e) => MediaCard.fromJson(e as Map<String, dynamic>)).toList(),
      episodes: (j['episodes'] as List? ?? const []).map((e) => EpisodeCard.fromJson(e as Map<String, dynamic>)).toList(),
      movies: (j['movies'] as List? ?? const []).map((e) => EpisodeCard.fromJson(e as Map<String, dynamic>)).toList(),
      audioTracks:
          (j['audioTracks'] as List? ?? const []).map((e) => DetailAudioTrack.fromJson(e as Map<String, dynamic>)).toList(),
      watch: j['watch'] != null ? MediaItemWatch.fromJson(j['watch'] as Map<String, dynamic>) : null,
      collections:
          (j['collections'] as List? ?? const []).map((e) => MediaCollection.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }
}

class MediaFileVideoInfo {
  final String? codec;
  final int? width;
  final int? height;
  final double? frameRate;
  final bool isHdr;
  MediaFileVideoInfo({required this.codec, required this.width, required this.height, required this.frameRate, required this.isHdr});
  factory MediaFileVideoInfo.fromJson(Map<String, dynamic> j) => MediaFileVideoInfo(
        codec: j['codec'] as String?,
        width: j['width'] as int?,
        height: j['height'] as int?,
        frameRate: (j['frameRate'] as num?)?.toDouble(),
        isHdr: j['isHdr'] as bool? ?? false,
      );
}

class MediaFileDescriptor {
  final String mediaFileId;
  final bool isPrimary;
  final String? container;
  final int? durationMs;
  final int? sizeBytes;
  final int? bitrate;
  final MediaFileVideoInfo? video;

  MediaFileDescriptor({
    required this.mediaFileId,
    required this.isPrimary,
    required this.container,
    required this.durationMs,
    required this.sizeBytes,
    required this.bitrate,
    required this.video,
  });

  factory MediaFileDescriptor.fromJson(Map<String, dynamic> j) => MediaFileDescriptor(
        mediaFileId: j['mediaFileId'] as String,
        isPrimary: j['isPrimary'] as bool? ?? false,
        container: j['container'] as String?,
        durationMs: j['durationMs'] as int?,
        sizeBytes: j['sizeBytes'] as int?,
        bitrate: j['bitrate'] as int?,
        video: j['video'] != null ? MediaFileVideoInfo.fromJson(j['video'] as Map<String, dynamic>) : null,
      );
}
